### ISSUE TITLE
cannot build on 'localhost': don't know how to open Nix store 'localhost'

### ISSUE THREAD

**Issue by basvandijk** [NixOS/nix#3177]
I just upgraded my Hydra from nixpkgs-19.03 to nixpkgs-19.09. 
This also upgraded Nix from 2.2.2 to 2.3. 

My Hydra has `localhost` configured as a builder which is/was needed to also run builds on the local Hydra machine:

```
# cat /etc/nix/machines | grep localhost
localhost x86_64-linux - 1 4 big-parallel
```

However now I see the following output in the log of a Hydra build when that build is  running on `localhost`:

```
cannot build on 'localhost': don't know how to open Nix store 'localhost'
```

After that error the log successfully continues. I also see that the build is running on one of my remote builders even though hydra thinks it's running on `localhost`.

What's the Nix-2.3 way of configuring hydra to also use `localhost` to build jobs?

**Comment by exarkun** [comment]
I have the same symptom using manual `nix-build` commands on a system configured with "localhost" as a buildMachine.

The Nix expression fragment configuring the system is:

```
    distributedBuilds = true;
    buildMachines = [
      { hostName = "saxtons";
        maxJobs = 12;
        speedFactor = 1;
        systems = [ "x86_64-linux" ];
      }
      { hostName = "localhost";
        maxJobs = 4;
        speedFactor = 1;
        systems = [ "x86_64-linux" ];
      }
    ];
```

This results in `/etc/nix/machines` containing:

```
saxtons x86_64-linux - 12 1
localhost x86_64-linux - 4 1
```

And when I run a `nix-build` command it produces output like this:

```
building '/nix/store/s22903k8rgfsja1cmgk4rxk1k2zf3npq-http-api-data-lib-http-api-data-0.4.1.drv' on 'ssh://saxtons'...                                                                       
building '/nix/store/zc7ixc0y9fqjzjggbr0f90s2xg78jdcs-vector-lib-vector-0.12.0.3.drv' on 'ssh://saxtons'...                                                                                  
cannot build on 'localhost': don't know how to open Nix store 'localhost'                                                                                                                    
building '/nix/store/kw9hlp9sngnncx42q276ss44k2kil4c6-word8-lib-word8-0.1.3.drv' on 'ssh://saxtons'...                                                                                       
cannot build on 'localhost': don't know how to open Nix store 'localhost'
building '/nix/store/8dh0hzqxbq3w8fywcfqy9vwq7wdhnd8d-x509-lib-x509-1.7.5.drv' on 'ssh://saxtons'...                                                                                         
cannot build on 'localhost': don't know how to open Nix store 'localhost'                                                                                                                    
building '/nix/store/ppvgxyryf7nwyclfwv10nhb02pwbi6cs-zlib-lib-zlib-0.6.2-config.drv' on 'ssh://saxtons'...                                                                                  
```

I am running NixOS 21.05, upgraded from an older NixOS install (original 20.09 perhaps).  I have nix and nix-daemon 2.3.11.

**Comment by regnat** [comment]
I think that’s an instance of https://github.com/NixOS/nix/issues/1874 . If that’s so, it’s fixed on master, and you can work around it by using the full `ssh://localhost` rather than `localhost`

**Comment by exarkun** [comment]
> I think that’s an instance of #1874 . If that’s so, it’s fixed on master, and you can work around it by using the full `ssh://localhost` rather than `localhost`

Hm.  It only affects "localhost" though.  Non-localhost builders still work fine.  Does it still seem like #1874?

**Comment by regnat** [comment]
> Hm. It only affects "localhost" though. Non-localhost builders still work fine. Does it still seem like #1874?

Yes it does, because #1874 only affects hostnames that start with `local`, `remote` or `autp` (https://github.com/NixOS/nix/pull/4938 has a bit more context actually). So it will break `localhost` but not `myotherhost`

**Comment by exarkun** [comment]
It doesn't seem like `ssh://localhost` (or `ssh://user@localhost`) is a viable workaround.  When /etc/nix/machines is configured with such a host, builds just hang waiting for a lock.  I assume because this defeats the detection of "localhost" meaning the build is actually local and special lock handling is needed?  I dunno.  But if it's fixed in master then I guess I can just upgrade after the next release.

**Comment by tomcur** [comment]
> It doesn't seem like `ssh://localhost` (or `ssh://user@localhost`) is a viable workaround. When /etc/nix/machines is configured with such a host, builds just hang waiting for a lock. I assume because this defeats the detection of "localhost" meaning the build is actually local and special lock handling is needed? I dunno. But if it's fixed in master then I guess I can just upgrade after the next release.

For now I think you'll be able to use `local` rather than `localhost` as a build machine host. This should special case to a local store. Perhaps you need to do `local?root=/nix/store`, I'm not entirely sure.

With `localhost` on master you'll encounter the same lock issue, as under the hood it'll fallback to SSH. In any case, there's a separate PR open on NixOS to enforce configured build machines to use SSH stores, as the current behavior of allowing special store types is unexpected. If there's a good reason to use special store types perhaps you can chime in: https://github.com/NixOS/nixpkgs/pull/127949.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
